### ISSUE TITLE
Install generator should place all views and assets in `/app`, not `/vendor`

### ISSUE THREAD

**Issue by jayroh** [thredded/thredded#159]
This relates to issue #157. Right now they end up in `/vendor` and should not.

**Comment by jayroh** [comment]
@blimey85 if you would like to handle this issue I could point you to the file in question right here :) :point_right: https://github.com/jayroh/thredded/blob/master/lib/generators/thredded/install/install_generator.rb

:smile:

**Comment by glebm** [comment]
@jayroh I don't think there should be a generator that copies everything! This makes updates very difficult, and is almost never ever necessary.

**Comment by glebm** [comment]
To clarify, I don't think there should be a generator that copies anything at all. Just proper instructions on overriding thredded views and styles.

**Comment by jayroh** [comment]
Before I say "I respectfully disagree, @glebm, my good sir :smile:" I want to make sure I'm understanding the approach you would suggest. Instead of a generator are you suggesting someone would have to manually find the files on their filesystem, in their gempath, and copy them over? Or, maybe download the raw file from github?

If that's the case then I can't say I agree. While I think that's doable, I think that's something we can make a heck of a lot easier with a generator (or part of a generator).

Is there a compromise, somewhere in between maybe? A flag for the respective parts of the UI, maybe? js, views, css, images, etc?

**Comment by glebm** [comment]
I was thinking along the lines of pointing to [Rails Guides: Overriding Engine Views](http://edgeguides.rubyonrails.org/engines.html#overriding-views), warning against doing this (noting that updates will become very difficult), and documenting a command like this:
```sh
# Copy the post view into the application to customize it:
mkdir -p app/views/thredded/posts && cp "$(bundle show thredded)/$_/_post.html.erb" "$_"
```

**Comment by blimey85** [comment]
What about how Devise does it, as an example?

Configuring views

We built Devise to help you quickly develop an application that uses authentication. However, we don't want to be in your way when you need to customize it.

Since Devise is an engine, all its views are packaged inside the gem. These views will help you get started, but after some time you may want to change them. If this is the case, you just need to invoke the following generator, and it will copy all views to your application:

rails generate devise:views
If you have more than one Devise model in your application (such as User and Admin), you will notice that Devise uses the same views for all models. Fortunately, Devise offers an easy way to customize views. All you need to do is setconfig.scoped_views = true inside the config/initializers/devise.rb file.

After doing so, you will be able to have views based on the role like users/sessions/new and admins/sessions/new. If no view is found within the scope, Devise will use the default view at devise/sessions/new. You can also use the generator to generate scoped views:

rails generate devise:views users
If you would like to generate only a few sets of views, like the ones for the registerable and confirmable module, you can pass a list of modules to the generator with the -v flag.

rails generate devise:views -v registrations confirmations

All views are in the gem, same as with Thredded, but if you want to bring them in, the generator can bring them in for you, to where they need to go. They also go a step(s) further and let you specify which views, and even do scoped views. While I’m comfortable on the command line I’m not sure everyone is going to want to be running commands to make directories and copy files over when a generator seems like it might be an easier option. Is it common that users would NOT need to edit templates to integrate with their site?

On March 22, 2016 at 6:36:48 PM, Gleb Mazovetskiy (notifications@github.com) wrote:

I was thinking along the lines of pointing to Rails Guides: Overriding Engine Views, warning against doing this (noting that updates will become very difficult), and documenting a command like this:

# Copy the post view into the application to customize it:
mkdir -p app/views/thredded/posts && cp "$(bundle show thredded)/$_/_post.html.erb" "$_"

**Comment by glebm** [comment]
If we decide to enable the users to copy all the views, generators are surely the way to go.

However, I think it's very dangerous to provide these, because it's a very bad idea to copy all the views:

1. Most users will only need to modify a couple of views.
2. They will run the generator, copying all of them.
3. On the next `bundle update`, they will forget to re-run the generator and inspect the diff, so they will:
    1. Miss out on features from thredded.
    2. Thredded will break for them and they will post an issue to the issue tracker.

In my experience, these issues are very hard to diagnose, e.g. just see the Devise tracker:
  * https://github.com/plataformatec/devise/search?q=regenerate+the+views&type=Issues&utf8=%E2%9C%93
  *  https://github.com/plataformatec/devise/search?utf8=%E2%9C%93&q=generate+the+views&type=Issues

**Comment by jayroh** [comment]
I agree that when people override the views or assets there are bound to be issues that creep up due to the evolution of the project. That's almost a given, to be honest. Full disclosure - this *just* happened with me and thoughtbot/administrate.

However, I think it's also a fact that *most* people implementing something like thredded will *not* want to stick with the stock UX/UI.

I do believe there's a happy medium here.

* Allow smaller subsets of files to be overridden, a la Devise's method per @blimey85's post above .. eg: `rails generate thredded:views -v topics` or `rails generate devise:stylesheets`, et al.
* Print out a nice reminder when they do this that they are now responsible for keeping the views in line with updates to the gem.
* Thoughts on a post-install message reminding people to update their customizations? Enh, maybe I'm not so much for this.

Would love some more feedback on this. @cgunther you've been working on integrating this gem quite a bit lately - what do you think? Pulling you in as well @TheDudeWithTheThing.

**Comment by cgunther** [comment]
Personally, I'd lean towards @glebm.

I've worked to do as much customizing as I can via CSS, so I require the thredded stylesheet, then define my own styles to adapt the out-of-the-box look to the site we're integrating thredded into.

On the HTML side, I've been very hesitant to override any templates/partials because I don't want to put myself in a scenario where I'm merging changes into my templates/partials from upstream thredded on every update. Luckily my requirements were quite flexible, so we were able to stick with thredded's out-of-the-box templates/partials in most cases. I've overriden 3 templates/partial so far:

1. `shared/_currently_online` - don't need this feature, so replaced it with a blank partial. Figured that was a little better than hiding with CSS, but maybe one day this could be a feature toggle.
2. `shared/_flash_messages` - since we're integrating into an existing application, we already have a pattern for showing flash message, so it doesn't need to be repeated within thredded (used an empty partial again). Maybe this could be smarter in the future to leave it up to the application to handle flash messages when thredded is not used as a standalone app (Devise leaves it up to the app).
3. `shared/_top_nav` - used this to add a big heading containing the messageboard name (or a generic title) above each page and don't need any of the standalone aspects.

In general, I like Devise's approach, but Devise's views seem much simpler compared to thredded, so I'm not sure if it's the greatest comparison. I likely wouldn't use a Devise-style generator because I want more finer grained control over which templates/partials I'm overriding, but maybe that's just me.

Having said all that, I think once you start customizing thredded (either via CSS or HTML), you're stuck paying close attention to the changes between versions and it's rarely as simple as a `bundle update`. Given the issues that can arise from customizing, it slightly feels fitting to me that you have to jump through more hoops, maybe that saves the customizing for the more experienced people.

**Comment by jayroh** [comment]
Those are some excellent points. Thank you @cgunther

**Comment by glebm** [comment]
Resolved with #178
